### PR TITLE
exclude replication from the concurrentUploadLimitMB

### PR DESCRIPTION
**Pull request by paochiang**

### CODE REVIEW

**Comment by kmlebedev** [comment]
@paochiang If you think waiting is better, but not safer, then add a time limit to the wait.

**Comment by kmlebedev** [comment]
@paochiang If you think waiting is better, but not safer, then add a time limit to the wait.

**Comment by kmlebedev** [comment]
@paochiang If you think waiting is better, but not safer, then add a time limit to the wait.

**Comment by paochiang** [comment]
@kmlebedev How about removing the client timeout? 
> Since waiting can easily lead to an infinite waiting loop on all instances. 

I think if cannot write the data successfully, the client should wait forever. All requests sent to the volume server should ensure that they are finally written successfully, and volume server doesn't need to care whether the client connection is closed.

**Comment by kmlebedev** [comment]
> @kmlebedev How about removing the client timeout?
> 
> > Since waiting can easily lead to an infinite waiting loop on all instances.
> 
> I think if cannot write the data successfully, the client should wait forever. All requests sent to the volume server should ensure that they are finally written successfully, and volume server doesn't need to care whether the client connection is closed.

Nobody forces the client to wait forever, and if something goes wrong, it is better to always quickly return the error, and the client himself will flexibly adapt and retry when he needs to and will wait as long as he needs

**Comment by kmlebedev** [comment]
There is also often a limit on the number of client connections on ingresses(nginx as HTTP load balancer)

**Comment by paochiang** [comment]
> @paochiang If you think waiting is better, but not safer, then add a time limit to the wait.

How long should this time limit?

**Comment by kmlebedev** [comment]
> > @paochiang If you think waiting is better, but not safer, then add a time limit to the wait.
> 
> How long should this time limit?

I think this limit should be configurable, but the default is given that the upstream http balancer waits for a response no more than 60 seconds

**Comment by paochiang** [comment]
@kmlebedev modified. I will test it in my env.

**Comment by kmlebedev** [comment]
> @kmlebedev modified. I will test it in my env.

lgtm

**Comment by paochiang** [comment]
> It is also worth checking on load tests to avoid #2755
> 
Verified locally, it's ok.

In addition, in my own environment, with new volume server, reading and writing operations look normal.
image

**Comment by kmlebedev** [comment]
> > It is also worth checking on load tests to avoid #2755
> 
> Verified locally, it's ok.
> 
> In addition, in my own environment, with new volume server, reading and writing operations look normal. image

Looks like testing concurrentDownloadLimit with get requests

**Comment by paochiang** [comment]
> Looks like testing concurrentDownloadLimit with get requests

It is an AI training task that reads and writes data normally, not testing `Concurrentdownloadlimit`

**Comment by chrislusf** [comment]
Prefer not to add a timer for each write request.

**Comment by paochiang** [comment]
modified,  change to lazy timeout check

File: weed/server/volume_server_handlers.go [hunk 1]
```diff
@@ -56,20 +57,33 @@ func (vs *VolumeServer) privateStoreHandler(w http.ResponseWriter, r *http.Reque
 		vs.guard.WhiteList(vs.DeleteHandler)(w, r)
 	case "PUT", "POST":
 
-		// wait until in flight data is less than the limit
 		contentLength := getContentLength(r)
-
+		startTime := time.Now()
+		vs.inFlightUploadDataLimitCond.L.Lock()
```
Comment (chrislusf) [inline weed/server/volume_server_handlers.go]: too many locks. The lock scope is also too big.

why not use the original atomic.AddInt64 operations?

Comment (paochiang) [reply weed/server/volume_server_handlers.go]: modified

**Comment by kmlebedev** [comment]
@paochiang How are things here?

**Comment by paochiang** [comment]
Waiting for review comments or the code to be merged. I don't know what else needs to be modified.

File: weed/server/volume_server_handlers.go [hunk 2]
```diff
@@ -56,20 +57,29 @@ func (vs *VolumeServer) privateStoreHandler(w http.ResponseWriter, r *http.Reque
...
-		if vs.concurrentUploadLimit != 0 && r.URL.Query().Get("type") != "replicate" &&
-			atomic.LoadInt64(&vs.inFlightUploadDataSize) > vs.concurrentUploadLimit {
-			err := fmt.Errorf("reject because inflight upload data %d > %d", vs.inFlightUploadDataSize, vs.concurrentUploadLimit)
-			glog.V(1).Infof("too many requests: %v", err)
-			writeJsonError(w, r, http.StatusTooManyRequests, err)
-			return
+		if r.URL.Query().Get("type") != "replicate" {
+			startTime := time.Now()
+			vs.inFlightUploadDataLimitCond.L.Lock()
+			for vs.concurrentUploadLimit != 0 && vs.inFlightUploadDataSize > vs.concurrentUploadLimit {
```
Comment (chrislusf) [inline weed/server/volume_server_handlers.go]: * move this `vs.concurrentUploadLimit != 0` out of the lock

Comment (paochiang) [reply weed/server/volume_server_handlers.go]: modified, moved to` if r.URL.Query().Get("type") != "replicate" {`  this line.

File: weed/server/volume_server_handlers.go [hunk 3]
```diff
@@ -56,20 +57,29 @@ func (vs *VolumeServer) privateStoreHandler(w http.ResponseWriter, r *http.Reque
...
+				}
+				glog.V(4).Infof("wait because inflight upload data %d > %d", vs.inFlightUploadDataSize, vs.concurrentUploadLimit)
+				vs.inFlightUploadDataLimitCond.Wait()
+			}
+			vs.inFlightUploadDataLimitCond.L.Unlock()
 		}
 		atomic.AddInt64(&vs.inFlightUploadDataSize, contentLength)
 		defer func() {
 			atomic.AddInt64(&vs.inFlightUploadDataSize, -contentLength)
+			vs.inFlightUploadDataLimitCond.Signal()
```
Comment (chrislusf) [inline weed/server/volume_server_handlers.go]: This also need to be conditional?

Comment (paochiang) [reply weed/server/volume_server_handlers.go]: It does not need.
image

Comment (chrislusf) [reply weed/server/volume_server_handlers.go]: It's still an unnecessary call.

Comment (paochiang) [reply weed/server/volume_server_handlers.go]: Got it.
`if r.URL.Query().Get("type") != "replicate" && vs.concurrentUploadLimit != 0 ` or `if vs.concurrentUploadLimit != 0`， which condition  is better?

File: weed/server/volume_server_handlers.go (hunk 2 above)
Comment (kmlebedev) [reply weed/server/volume_server_handlers.go]: It seems that Or logic is needed

Comment (paochiang) [reply weed/server/volume_server_handlers.go]: > It seems that Or logic is needed

`if r.URL.Query().Get("type") != "replicate" || vs.concurrentUploadLimit != 0 {` ? 
 It cannot exclude replication writing. It should be `and` logic.

File: weed/server/volume_server_handlers.go (hunk 3 above)
Comment (paochiang) [reply weed/server/volume_server_handlers.go]: modified, added `if vs.concurrentUploadLimit != 0 {`

File: weed/server/volume_server_handlers.go (hunk 2 above)
Comment (kmlebedev) [reply weed/server/volume_server_handlers.go]: ok

**Comment by kmlebedev** [comment]
This must be checked on load tests with replication. Set myself a task

**Pull request merged**